src/unix/pipe.c

```diff
@@ -29,6 +29,40 @@
 #include <unistd.h>
 #include <stdlib.h>
 
+struct uv__sockaddr_un_path {
+  char pad[offsetof(struct sockaddr_un, sun_path)];
+#ifdef SOCK_MAXADDRLEN
+  /* macOS and the BSDs support paths > sizeof(su.sun_path).
+   * SOCK_MAXADDRLEN is, to the best of my knowledge, always 255.
+   * Minus the two byte header, that leaves 253 bytes for the path.
+   */
+  char path[SOCK_MAXADDRLEN - offsetof(struct sockaddr_un, sun_path)];
+#else
+  char path[sizeof(((struct sockaddr_un*)0)->sun_path)];
+#endif
+};
+
+union uv__sockaddr_un {
+  struct sockaddr sa;
+  struct uv__sockaddr_un_path up;
+};
```

Comment:
Is `offsetof` a legal constexpr for C89? I'm not sure the type-punning (via the union) is desirable either, as it's not legal C code, though we are intending to turn off TBAA.

I wonder if we should instead use the normal struct definition, then add a storage type with the extra padding:
```suggestion
union uv__sockaddr_un_storage {
  struct sockaddr_un sa;
#ifdef SOCK_MAXADDRLEN
   /* macOS and the BSDs support paths > sizeof(su.sun_path).
    * SOCK_MAXADDRLEN is, to the best of my knowledge, always 255.
    * Minus the two byte header, that leaves 253 bytes for the path.
    */
  char up[SOCK_MAXADDRLEN];
#else
  char up[sizeof(struct sockaddr_un)];
#endif
};
```
elsewhere, then, instead of `sizeof(s->up.path)` then, it'd be either `sizeof(s->sa.path) + sizeof(s-> up) - sizeof(s->sa)` or equivalently `sizeof(s->up) - offsetof(struct sockaddr_un, sun_path)`

Reply:
Sorry @vtjnash, missed your comment.

> Is offsetof a legal constexpr for C89?

Yes, IMO (compiler builtin), and easy enough to fix if someone somewhere is using an archaic compiler.

> I'm not sure the type-punning (via the union) is desirable either

I suspect that's an academic concern because the only aliasing is between `struct sockaddr` and `char pad[]`. Even if it's not theoretic, libuv is built with `-fno-strict-aliasing` now.

Your suggestion would work but it feels like it has a larger margin for error so I'd rather stick with my approach.